File: backend/app/routers/reports.py
"""Sales, profit, inventory and cashier reporting."""
import csv
import io
from datetime import date, timedelta

from fastapi import APIRouter, Depends, Query
from fastapi.responses import StreamingResponse

from .. import db
from ..security import require
from ..services import analytics

router = APIRouter(prefix="/api/reports", tags=["reports"])


def _range(date_from, date_to, default_days=30):
    end = date.fromisoformat(date_to) if date_to else date.today()
    start = date.fromisoformat(date_from) if date_from else end - timedelta(days=default_days - 1)
    return start.isoformat(), end.isoformat()


@router.get("/dashboard")
def dashboard(user=Depends(require("reports.view"))):
    """Today / this month at a glance, plus a 30-day trend line."""
    today = date.today().isoformat()
    month_start = date.today().replace(day=1).isoformat()

    def totals(clause, params):
        row = db.query_one(
            f"""SELECT COUNT(*) AS receipts,
                       COALESCE(SUM(total), 0) AS sales,
                       COALESCE(SUM(parts_total), 0) AS parts,
                       COALESCE(SUM(labor_total), 0) AS labor,
                       COALESCE(SUM(profit), 0) AS profit,
                       COALESCE(SUM(discount), 0) AS discount
                  FROM sales WHERE status = 'completed' AND {clause}""",
            params,
        )
        data = {k: round(float(row[k] or 0), 2) for k in row.keys()}
        data["receipts"] = int(row["receipts"])
        data["average_sale"] = round(data["sales"] / data["receipts"], 2) if data["receipts"] else 0
        return data

    trend = db.query(
        """SELECT business_date AS d, SUM(total) AS sales, SUM(profit) AS profit,
                  COUNT(*) AS receipts
             FROM sales
            WHERE status = 'completed' AND business_date >= date('now', '-29 days')
         GROUP BY business_date ORDER BY business_date"""
    )
    stock = db.query_one(
        """SELECT COUNT(*) AS skus, COALESCE(SUM(stock_qty * unit_cost), 0) AS cost_value,
                  COALESCE(SUM(stock_qty * retail_price), 0) AS retail_value,
                  COALESCE(SUM(CASE WHEN stock_qty <= 0 THEN 1 ELSE 0 END), 0) AS out_of_stock,
                  COALESCE(SUM(CASE WHEN stock_qty <= reorder_point THEN 1 ELSE 0 END), 0) AS critical
             FROM items WHERE active = 1 AND delisted = 0"""
    )
    return {
        "today": totals("business_date = ?", (today,)),
        "month": totals("business_date >= ?", (month_start,)),
        "stock": {
            "skus": int(stock["skus"]),
            "cost_value": round(float(stock["cost_value"]), 2),
            "retail_value": round(float(stock["retail_value"]), 2),
            "potential_margin": round(
                float(stock["retail_value"]) - float(stock["cost_value"]), 2
            ),
            "out_of_stock": int(stock["out_of_stock"]),
            "critical": int(stock["critical"]),
        },
        "trend": [
            {"date": r["d"], "sales": round(float(r["sales"]), 2),
             "profit": round(float(r["profit"]), 2), "receipts": r["receipts"]}
            for r in trend
        ],
    }


@router.get("/sales-summary")
def sales_summary(
    date_from: str | None = None,
    date_to: str | None = None,
    group_by: str = Query(default="day", pattern="^(day|week|month)$"),
    user=Depends(require("reports.view")),
):
    start, end = _range(date_from, date_to)
    bucket = {
        "day": "business_date",
        "week": "strftime('%Y-W%W', business_date)",
        "month": "strftime('%Y-%m', business_date)",
    }[group_by]

    rows = db.query(
        f"""SELECT {bucket} AS period, COUNT(*) AS receipts,
                   SUM(total) AS sales, SUM(parts_total) AS parts, SUM(labor_total) AS labor,
                   SUM(cost_total) AS cost, SUM(profit) AS profit, SUM(discount) AS discount
              FROM sales
             WHERE status = 'completed' AND business_date BETWEEN ? AND ?
          GROUP BY period ORDER BY period""",
        (start, end),
    )
    periods = [
        {
            "period": r["period"], "receipts": r["receipts"],
            "sales": round(float(r["sales"] or 0), 2),
            "parts": round(float(r["parts"] or 0), 2),
            "labor": round(float(r["labor"] or 0), 2),
            "cost": round(float(r["cost"] or 0), 2),
            "profit": round(float(r["profit"] or 0), 2),
            "discount": round(float(r["discount"] or 0), 2),
            "margin_pct": round(float(r["profit"] or 0) / float(r["sales"]) * 100, 1)
                          if float(r["sales"] or 0) > 0 else 0.0,
        }
        for r in rows
    ]
    totals = {
        key: round(sum(p[key] for p in periods), 2)
        for key in ("sales", "parts", "labor", "cost", "profit", "discount")
    }
    totals["receipts"] = sum(p["receipts"] for p in periods)
    totals["margin_pct"] = (
        round(totals["profit"] / totals["sales"] * 100, 1) if totals["sales"] > 0 else 0.0
    )
    return {"range": {"from": start, "to": end}, "group_by": group_by,
            "periods": periods, "totals": totals}


@router.get("/top-items")
def top_items(
    date_from: str | None = None,
    date_to: str | None = None,
    by: str = Query(default="revenue", pattern="^(revenue|qty|profit)$"),
    limit: int = Query(default=25, le=200),
    user=Depends(require("reports.view")),
):
    start, end = _range(date_from, date_to)
    order = {"revenue": "revenue", "qty": "qty", "profit": "profit"}[by]
    rows = db.query(
        f"""SELECT l.item_id, l.sku, l.description, l.line_type,
                   SUM(l.qty) AS qty, SUM(l.total) AS revenue, SUM(l.profit) AS profit
              FROM sale_lines l JOIN sales s ON s.id = l.sale_id
             WHERE s.status = 'completed' AND s.business_date BETWEEN ? AND ?
          GROUP BY l.item_id, l.service_id, l.description
          ORDER BY {order} DESC LIMIT ?""",
        (start, end, limit),
    )
    return {
        "range": {"from": start, "to": end},
        "by": by,
        "items": [
            {**dict(r), "qty": round(float(r["qty"]), 2),
             "revenue": round(float(r["revenue"]), 2),
             "profit": round(float(r["profit"]), 2)}
            for r in rows
        ],
    }


@router.get("/category-performance")
def category_performance(date_from: str | None = None, date_to: str | None = None,
                         user=Depends(require("reports.view"))):
    start, end = _range(date_from, date_to)
    rows = db.query(
        """SELECT COALESCE(c.name, 'UNCATEGORISED') AS category,
                  SUM(l.qty) AS qty, SUM(l.total) AS revenue, SUM(l.profit) AS profit
             FROM sale_lines l
             JOIN sales s ON s.id = l.sale_id
             LEFT JOIN items i ON i.id = l.item_id
             LEFT JOIN categories c ON c.id = i.category_id
            WHERE s.status = 'completed' AND l.line_type = 'item'
              AND s.business_date BETWEEN ? AND ?
         GROUP BY category ORDER BY revenue DESC""",
        (start, end),
    )
    stock = {
        r["category"]: round(float(r["value"] or 0), 2)
        for r in db.query(
            """SELECT COALESCE(c.name, 'UNCATEGORISED') AS category,
                      SUM(i.stock_qty * i.unit_cost) AS value
                 FROM items i LEFT JOIN categories c ON c.id = i.category_id
                WHERE i.active = 1 AND i.delisted = 0 GROUP BY category"""
        )
    }
    return {
        "range": {"from": start, "to": end},
        "categories": [
            {
                "category": r["category"],
                "qty": round(float(r["qty"] or 0), 2),
                "revenue": round(float(r["revenue"] or 0), 2),
                "profit": round(float(r["profit"] or 0), 2),
                "stock_value": stock.get(r["category"], 0.0),
            }
            for r in rows
        ],
    }


@router.get("/inventory-valuation")
def inventory_valuation(user=Depends(require("reports.financial"))):
    rows = db.query(
        """SELECT COALESCE(c.name, 'UNCATEGORISED') AS category,
                  COUNT(*) AS skus, SUM(i.stock_qty) AS units,
                  SUM(i.stock_qty * i.unit_cost) AS cost_value,
                  SUM(i.stock_qty * i.retail_price) AS retail_value
             FROM items i LEFT JOIN categories c ON c.id = i.category_id
            WHERE i.active = 1 AND i.delisted = 0
         GROUP BY category ORDER BY cost_value DESC"""
    )
    categories = [
        {
            "category": r["category"], "skus": r["skus"],
            "units": round(float(r["units"] or 0), 2),
            "cost_value": round(float(r["cost_value"] or 0), 2),
            "retail_value": round(float(r["retail_value"] or 0), 2),
            "potential_margin": round(
                float(r["retail_value"] or 0) - float(r["cost_value"] or 0), 2
            ),
        }
        for r in rows
    ]
    return {
        "categories": categories,
        "totals": {
            "skus": sum(c["skus"] for c in categories),
            "units": round(sum(c["units"] for c in categories), 2),
            "cost_value": round(sum(c["cost_value"] for c in categories), 2),
            "retail_value": round(sum(c["retail_value"] for c in categories), 2),
            "potential_margin": round(sum(c["potential_margin"] for c in categories), 2),
        },
    }


@router.get("/profit-and-loss")
def profit_and_loss(date_from: str | None = None, date_to: str | None = None,
                    user=Depends(require("reports.financial"))):
    """Trading account for the period: parts margin plus labour income."""
    start, end = _range(date_from, date_to)
    row = db.query_one(
        """SELECT COALESCE(SUM(parts_total), 0) AS parts,
                  COALESCE(SUM(labor_total), 0) AS labor,
                  COALESCE(SUM(cost_total), 0) AS cogs,
                  COALESCE(SUM(discount), 0) AS discounts,
                  COALESCE(SUM(total), 0) AS net_sales,
                  COUNT(*) AS receipts
             FROM sales WHERE status = 'completed' AND business_date BETWEEN ? AND ?""",
        (start, end),
    )
    parts = round(float(row["parts"]), 2)
    labor = round(float(row["labor"]), 2)
    cogs = round(float(row["cogs"]), 2)
    net_sales = round(float(row["net_sales"]), 2)
    gross_profit = round(parts - cogs, 2)

    purchases = db.query_one(
        """SELECT COALESCE(SUM(m.qty_delta * m.unit_cost), 0) AS v
             FROM stock_moves m
            WHERE m.move_type = 'purchase' AND date(m.ts) BETWEEN ? AND ?""",
        (start, end),
    )["v"]

    return {
        "range": {"from": start, "to": end},
        "receipts": int(row["receipts"]),
        "parts_sales": parts,
        "service_income": labor,
        "discounts_given": round(float(row["discounts"]), 2),
        "net_sales": net_sales,
        "cost_of_goods_sold": cogs,
        "gross_profit_on_parts": gross_profit,
        "total_gross_profit": round(gross_profit + labor, 2),
        "gross_margin_pct": round((gross_profit + labor) / net_sales * 100, 1)
                            if net_sales > 0 else 0.0,
        "stock_purchased": round(float(purchases), 2),
    }


@router.get("/cashier")
def cashier_report(date_from: str | None = None, date_to: str | None = None,
                   user=Depends(require("reports.view"))):
    """Per-cashier takings and tender breakdown — the shift (X/Z) read."""
    start, end = _range(date_from, date_to, default_days=1)
    rows = db.query(
        """SELECT u.id, u.username, u.full_name, COUNT(*) AS receipts,
                  SUM(s.total) AS sales, SUM(s.discount) AS discounts, SUM(s.profit) AS profit
             FROM sales s JOIN users u ON u.id = s.user_id
            WHERE s.status = 'completed' AND s.business_date BETWEEN ? AND ?
         GROUP BY u.id ORDER BY sales DESC""",
        (start, end),
    )
    tenders = db.query(
        """SELECT p.method, SUM(p.amount) AS amount, COUNT(*) AS count
             FROM payments p JOIN sales s ON s.id = p.sale_id
            WHERE s.status = 'completed' AND s.business_date BETWEEN ? AND ?
         GROUP BY p.method ORDER BY amount DESC""",
        (start, end),
    )
    voids = db.query(
        """SELECT s.receipt_no, s.total, s.void_reason, u.username AS voided_by, s.voided_at
             FROM sales s LEFT JOIN users u ON u.id = s.voided_by
            WHERE s.status = 'voided' AND s.business_date BETWEEN ? AND ?
         ORDER BY s.id DESC""",
        (start, end),
    )
    return {
        "range": {"from": start, "to": end},
        "cashiers": [
            {**dict(r), "sales": round(float(r["sales"] or 0), 2),
             "discounts": round(float(r["discounts"] or 0), 2),
             "profit": round(float(r["profit"] or 0), 2)}
            for r in rows
        ],
        "tenders": [
            {"method": t["method"], "amount": round(float(t["amount"]), 2),
             "count": t["count"]}
            for t in tenders
        ],
        "voids": [dict(v) for v in voids],
    }


@router.get("/export/{dataset}")
def export_csv(
    dataset: str,
    date_from: str | None = None,
    date_to: str | None = None,
    days: int = 90,
    user=Depends(require("reports.view")),
):
    """Download any core dataset as CSV for Excel or the accountant."""
    buffer = io.StringIO()
    writer = csv.writer(buffer)

    if dataset == "inventory":
        writer.writerow(["SKU", "Description", "Category", "Supplier", "Unit Cost",
                         "Retail Price", "Stock Qty", "Stock Value", "Reorder Point"])
        for r in db.query(
            """SELECT i.sku, i.description, COALESCE(c.name,'') AS category,
                      COALESCE(s.code,'') AS supplier, i.unit_cost, i.retail_price,
                      i.stock_qty, i.reorder_point
                 FROM items i
                 LEFT JOIN categories c ON c.id = i.category_id
                 LEFT JOIN suppliers s ON s.id = i.supplier_id
                WHERE i.active = 1 AND i.delisted = 0 ORDER BY i.description"""
        ):
            writer.writerow([
                r["sku"], r["description"], r["category"], r["supplier"],
                r["unit_cost"], r["retail_price"], r["stock_qty"],
                round(float(r["stock_qty"]) * float(r["unit_cost"]), 2), r["reorder_point"],
            ])

    elif dataset == "sales":
        start, end = _range(date_from, date_to)
        writer.writerow(["Receipt", "Date", "Cashier", "Customer", "Item Code",
                         "Description", "Qty", "Price", "Discount", "Total", "Profit"])
        for r in db.query(
            """SELECT s.receipt_no, s.business_date, u.username, s.customer_name,
                      l.sku, l.description, l.qty, l.unit_price, l.discount, l.total, l.profit
                 FROM sale_lines l
                 JOIN sales s ON s.id = l.sale_id
                 JOIN users u ON u.id = s.user_id
                WHERE s.status = 'completed' AND s.business_date BETWEEN ? AND ?
             ORDER BY s.id, l.id""",
            (start, end),
        ):
            writer.writerow(list(r))

    elif dataset == "reorder":
        rows, _, _ = analytics.reorder_suggestions(days=days)
        writer.writerow(["Urgency", "SKU", "Description", "Supplier", "On Hand", "On Order",
                         "Reorder Point", "Suggested Qty", "Unit Cost", "Order Cost",
                         "Days Cover", "Reason"])
        for r in rows:
            writer.writerow([
                r["urgency"], r["sku"], r["description"], r["supplier"], r["on_hand"],
                r["on_order"], r["reorder_point"], r["suggested_qty"], r["unit_cost"],
                r["order_cost"], r["days_of_cover"], r["reason"],
            ])

    elif dataset == "movers":
        rows, _, _ = analytics.analyze(days=days)
        rows.sort(key=lambda r: -r["daily_rate"])
        writer.writerow(["SKU", "Description", "Category", "Movement", "ABC-XYZ",
                         "Sold Qty", "Revenue", "Gross Profit", "Monthly Forecast",
                         "On Hand", "Days Cover"])
        for r in rows:
            writer.writerow([
                r["sku"], r["description"], r["category"], r["movement"], r["abc_xyz"],
                r["sold_qty"], r["revenue"], r["gross_profit"], r["monthly_rate"],
                r["on_hand"], r["days_of_cover"],
            ])
    else:
        from fastapi import HTTPException
        raise HTTPException(
            status_code=404,
            detail="Unknown dataset. Use inventory, sales, reorder or movers.",
        )

    buffer.seek(0)
    filename = f"kygs-{dataset}-{date.today().isoformat()}.csv"
    return StreamingResponse(
        iter([buffer.getvalue()]),
        media_type="text/csv",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )

File: frontend/js/views/reports.js
// Sales, profit, category and cashier reporting.

import { api, session } from '../api.js';
import {
  badge, barChart, daysAgo, empty, esc, loading, money, num, pct,
  statTile, table, toast, today,
} from '../ui.js';

const state = { tab: 'sales', from: daysAgo(29), to: today(), group_by: 'day' };

export async function render(root) {
  root.innerHTML = `
    <div class="page-head">
      <div><h2>Reports</h2><p>Trading performance, margins and till accountability.</p></div>
      <div class="page-actions">
        <label class="field" style="margin:0"><span>From</span>
          <input type="date" id="r-from" value="${state.from}"></label>
        <label class="field" style="margin:0"><span>To</span>
          <input type="date" id="r-to" value="${state.to}"></label>
      </div>
    </div>

    <div class="tab-row">
      <button class="tab active" data-tab="sales">Sales</button>
      <button class="tab" data-tab="items">Top items</button>
      <button class="tab" data-tab="categories">Categories</button>
      <button class="tab" data-tab="cashier">Cashiers &amp; tenders</button>
      ${session.can('reports.financial')
        ? '<button class="tab" data-tab="pnl">Profit &amp; loss</button>' : ''}
      ${session.can('reports.financial')
        ? '<button class="tab" data-tab="valuation">Stock valuation</button>' : ''}
      <button class="tab" data-tab="receipts">Receipts</button>
    </div>

    <div id="r-body">${loading()}</div>`;

  document.getElementById('r-from').addEventListener('change', (event) => {
    state.from = event.target.value; load();
  });
  document.getElementById('r-to').addEventListener('change', (event) => {
    state.to = event.target.value; load();
  });
  root.querySelectorAll('[data-tab]').forEach((tab) => {
    tab.addEventListener('click', () => {
      state.tab = tab.dataset.tab;
      root.querySelectorAll('[data-tab]').forEach((t) => t.classList.toggle('active', t === tab));
      load();
    });
  });

  load();
}

const range = () => ({ date_from: state.from, date_to: state.to });

async function load() {
  const box = document.getElementById('r-body');
  box.innerHTML = loading();
  try {
    const views = {
      sales: salesView, items: itemsView, categories: categoriesView,
      cashier: cashierView, pnl: pnlView, valuation: valuationView, receipts: receiptsView,
    };
    box.innerHTML = await views[state.tab]();
    wire();
  } catch (error) {
    box.innerHTML = `<div class="alert alert-error">${esc(error.message)}</div>`;
  }
}

function wire() {
  document.getElementById('r-group')?.addEventListener('change', (event) => {
    state.group_by = event.target.value; load();
  });
  document.getElementById('r-export')?.addEventListener('click', () =>
    api.download('/api/reports/export/sales', range()));
  document.querySelectorAll('[data-receipt]').forEach((row) => {
    row.addEventListener('click', () => openReceipt(Number(row.dataset.receipt)));
  });
}

async function salesView() {
  const data = await api.get('/api/reports/sales-summary', {
    ...range(), group_by: state.group_by,
  });
  const t = data.totals;

  return `
    <div class="grid grid-4">
      ${statTile({ label: 'Net sales', value: money(t.sales), tone: 'accent',
                   sub: `${num(t.receipts)} receipts` })}
      ${statTile({ label: 'Gross profit', value: money(t.profit), tone: 'ok',
                   sub: `${pct(t.margin_pct)} margin` })}
      ${statTile({ label: 'Parts vs labour', value: money(t.parts), tone: 'info',
                   sub: `labour ${money(t.labor)}` })}
      ${statTile({ label: 'Discounts given', value: money(t.discount),
                   tone: t.discount > 0 ? 'warn' : '' })}
    </div>

    <div class="card" style="margin-top:16px">
      <div class="card-head"><h3>Sales over time</h3>
        <div style="display:flex;gap:8px;align-items:center">
          <select id="r-group" style="width:auto">
            <option value="day" ${state.group_by === 'day' ? 'selected' : ''}>By day</option>
            <option value="week" ${state.group_by === 'week' ? 'selected' : ''}>By week</option>
            <option value="month" ${state.group_by === 'month' ? 'selected' : ''}>By month</option>
          </select>
          <button class="btn btn-sm" id="r-export">Export CSV</button>
        </div></div>
      ${data.periods.length
        ? barChart(data.periods.map((p) => ({ label: p.period, value: p.sales })),
                   { format: money })
        : empty('No sales in this period')}
    </div>

    <div class="card">
      ${table([
        { label: 'Period', key: 'period', nowrap: true },
        { label: 'Receipts', align: 'right', render: (r) => num(r.receipts) },
        { label: 'Parts', align: 'right', render: (r) => money(r.parts) },
        { label: 'Labour', align: 'right', render: (r) => money(r.labor) },
        { label: 'Discount', align: 'right', render: (r) => money(r.discount) },
        { label: 'Net sales', align: 'right', render: (r) => `<strong>${money(r.sales)}</strong>` },
        { label: 'Cost', align: 'right', render: (r) => money(r.cost) },
        { label: 'Profit', align: 'right', render: (r) =>
            `<span style="color:${r.profit >= 0 ? 'var(--ok)' : 'var(--danger)'}">
               ${money(r.profit)}</span>` },
        { label: 'Margin', align: 'right', render: (r) => pct(r.margin_pct) },
      ], data.periods, { emptyMessage: 'No sales in this period' })}
    </div>`;
}

async function itemsView() {
  const [byRevenue, byQty] = await Promise.all([
    api.get('/api/reports/top-items', { ...range(), by: 'revenue', limit: 25 }),
    api.get('/api/reports/top-items', { ...range(), by: 'qty', limit: 25 }),
  ]);

  const columns = [
    { label: 'Item', render: (r) => `<strong>${esc(r.description)}</strong>
        <div class="faint" style="font-size:11.5px">${esc(r.sku || r.line_type)}</div>` },
    { label: 'Qty', align: 'right', render: (r) => num(r.qty) },
    { label: 'Revenue', align: 'right', render: (r) => money(r.revenue) },
    { label: 'Profit', align: 'right', render: (r) => money(r.profit) },
  ];

  return `
    <div class="grid grid-2">
      <div class="card">
        <div class="card-head"><h3>Top earners</h3><span class="hint">by revenue</span></div>
        ${table(columns, byRevenue.items)}
      </div>
      <div class="card">
        <div class="card-head"><h3>Highest volume</h3><span class="hint">by units sold</span></div>
        ${table(columns, byQty.items)}
      </div>
    </div>`;
}

async function categoriesView() {
  const data = await api.get('/api/reports/category-performance', range());
  return `
    <div class="card">
      <div class="card-head"><h3>Category performance</h3></div>
      ${data.categories.length
        ? barChart(data.categories.map((c) => ({ label: c.category, value: c.revenue })),
                   { format: money })
        : ''}
      ${table([
        { label: 'Category', render: (r) => `<strong>${esc(r.category)}</strong>` },
        { label: 'Units', align: 'right', render: (r) => num(r.qty) },
        { label: 'Revenue', align: 'right', render: (r) => money(r.revenue) },
        { label: 'Profit', align: 'right', render: (r) => money(r.profit) },
        { label: 'Margin', align: 'right', render: (r) =>
            pct(r.revenue > 0 ? (r.profit / r.revenue) * 100 : 0) },
        { label: 'Stock value', align: 'right', render: (r) => money(r.stock_value) },
      ], data.categories, { emptyMessage: 'No category sales in this period' })}
    </div>`;
}

async function cashierView() {
  const data = await api.get('/api/reports/cashier', range());
  return `
    <div class="grid grid-2">
      <div class="card">
        <div class="card-head"><h3>By cashier</h3></div>
        ${table([
          { label: 'Cashier', render: (r) => `<strong>${esc(r.username)}</strong>
              <div class="faint" style="font-size:11.5px">${esc(r.full_name || '')}</div>` },
          { label: 'Receipts', align: 'right', render: (r) => num(r.receipts) },
          { label: 'Sales', align: 'right', render: (r) => money(r.sales) },
          { label: 'Discounts', align: 'right', render: (r) => money(r.discounts) },
          { label: 'Profit', align: 'right', render: (r) => money(r.profit) },
        ], data.cashiers, { emptyMessage: 'No sales in this period' })}
      </div>
      <div class="card">
        <div class="card-head"><h3>Tenders taken</h3></div>
        ${table([
          { label: 'Method', render: (r) => badge(r.method, 'medium') },
          { label: 'Count', align: 'right', render: (r) => num(r.count) },
          { label: 'Amount', align: 'right', render: (r) => money(r.amount) },
        ], data.tenders, { emptyMessage: 'No payments recorded' })}
      </div>
    </div>
    <div class="card">
      <div class="card-head"><h3>Voided sales</h3>
        <span class="hint">every void is logged against the user who made it</span></div>
      ${table([
        { label: 'Receipt', render: (r) => esc(r.receipt_no), nowrap: true },
        { label: 'Value', align: 'right', render: (r) => money(r.total) },
        { label: 'Voided by', render: (r) => esc(r.voided_by || '—') },
        { label: 'When', render: (r) => esc(String(r.voided_at || '').slice(0, 16)), nowrap: true },
        { label: 'Reason', render: (r) => `<span class="faint">${esc(r.void_reason)}</span>` },
      ], data.voids, { emptyMessage: 'No voids in this period — good' })}
    </div>`;
}

async function pnlView() {
  const data = await api.get('/api/reports/profit-and-loss', range());
  return `
    <div class="grid grid-4">
      ${statTile({ label: 'Net sales', value: money(data.net_sales), tone: 'accent',
                   sub: `${num(data.receipts)} receipts` })}
      ${statTile({ label: 'Total gross profit', value: money(data.total_gross_profit),
                   tone: 'ok', sub: `${pct(data.gross_margin_pct)} margin` })}
      ${statTile({ label: 'Service income', value: money(data.service_income), tone: 'info',
                   sub: 'labour, no cost of goods' })}
      ${statTile({ label: 'Stock purchased', value: money(data.stock_purchased),
                   sub: 'deliveries booked in' })}
    </div>
    <div class="card" style="margin-top:16px">
      <div class="card-head"><h3>Trading account
        <span class="muted" style="font-weight:400">
          ${esc(data.range.from)} → ${esc(data.range.to)}</span></h3></div>
      <dl class="kv" style="max-width:520px">
        <dt>Parts sales</dt><dd>${money(data.parts_sales)}</dd>
        <dt>Service income</dt><dd>${money(data.service_income)}</dd>
        <dt>Less: discounts given</dt><dd>−${money(data.discounts_given)}</dd>
        <dt class="strong">Net sales</dt><dd class="strong">${money(data.net_sales)}</dd>
        <dt>Less: cost of goods sold</dt><dd>−${money(data.cost_of_goods_sold)}</dd>
        <dt>Gross profit on parts</dt><dd>${money(data.gross_profit_on_parts)}</dd>
        <dt class="strong">Total gross profit</dt>
        <dd class="strong">${money(data.total_gross_profit)}</dd>
      </dl>
      <p class="faint" style="margin-top:14px;font-size:12.5px">
        Operating expenses (rent, wages, utilities) are not tracked by the till, so this is a
        trading account rather than a full profit and loss statement.</p>
    </div>`;
}

async function valuationView() {
  const data = await api.get('/api/reports/inventory-valuation');
  const t = data.totals;
  return `
    <div class="grid grid-4">
      ${statTile({ label: 'Stock at cost', value: money(t.cost_value), tone: 'accent',
                   sub: `${num(t.skus)} SKUs · ${num(t.units)} units` })}
      ${statTile({ label: 'Stock at retail', value: money(t.retail_value), tone: 'info' })}
      ${statTile({ label: 'Potential margin', value: money(t.potential_margin), tone: 'ok',
                   sub: 'if everything sold at list price' })}
      ${statTile({ label: 'Average cost per SKU',
                   value: money(t.skus ? t.cost_value / t.skus : 0) })}
    </div>
    <div class="card" style="margin-top:16px">
      <div class="card-head"><h3>Valuation by category</h3></div>
      ${table([
        { label: 'Category', render: (r) => `<strong>${esc(r.category)}</strong>` },
        { label: 'SKUs', align: 'right', render: (r) => num(r.skus) },
        { label: 'Units', align: 'right', render: (r) => num(r.units) },
        { label: 'At cost', align: 'right', render: (r) => money(r.cost_value) },
        { label: 'At retail', align: 'right', render: (r) => money(r.retail_value) },
        { label: 'Potential margin', align: 'right', render: (r) => money(r.potential_margin) },
      ], data.categories)}
    </div>`;
}

async function receiptsView() {
  const data = await api.get('/api/pos/sales', { ...range(), limit: 300 });
  return `
    <div class="card">
      <div class="card-head"><h3>Receipts</h3>
        <span class="hint">${num(data.total)} in this period — click to reprint</span></div>
      ${table([
        { label: 'Receipt', render: (r) => `<strong>${esc(r.receipt_no)}</strong>`, nowrap: true },
        { label: 'When', render: (r) => esc(String(r.ts).slice(0, 16)), nowrap: true },
        { label: 'Cashier', render: (r) => esc(r.cashier) },
        { label: 'Customer', render: (r) => esc(r.customer_name || '—') },
        { label: 'Total', align: 'right', render: (r) => money(r.total) },
        { label: 'Profit', align: 'right', render: (r) => money(r.profit) },
        { label: 'Status', render: (r) =>
            badge(r.status, r.status === 'voided' ? 'danger' : 'ok') },
      ], data.sales, {
        emptyMessage: 'No receipts in this period',
        rowAttrs: (r) => `data-receipt="${r.id}"`,
        onRowClick: true,
      })}
    </div>`;
}

async function openReceipt(saleId) {
  try {
    const payload = await api.get(`/api/pos/sales/${saleId}`);
    const { showReceipt } = await import('./pos.js');
    showReceipt(payload);
  } catch (error) {
    toast(error.message, 'error');
  }
}


